src/main/scala/com/linkedin/spark/datasources/tfrecord/TFRecordFileReader.scala

```diff
@@ -35,9 +35,9 @@ object TFRecordFileReader {
     // Ensure that the reader is closed even if the task fails or doesn't consume the entire
     // iterator of records.
     Option(TaskContext.get()).foreach { taskContext =>
-      taskContext.addTaskCompletionListener { _ =>
+      taskContext.addTaskCompletionListener[Unit]((_: TaskContext) =>
```

Comment:
Is this required by Spark 2.4.6?

Reply:
This isn't strictly required for Spark 2.4.6 (if built with Scala 2.11), but it is required for Spark 3.0.0 since the only version of Scala that Spark 3.0.0 is built with is Scala 2.12. There was a similar issue which was fixed in the same manner in the Spark repo, see https://github.com/apache/spark/pull/22229

Reply:
This change would cause build failure with spark 2.3.0 `mvn clean install -Dspark.version=2.3.0`:
```
[ERROR] spark-tfrecord/src/main/scala/com/linkedin/spark/datasources/tfrecord/TFRecordFileReader.scala:38: overloaded method value addTaskCompletionListener with alternatives:
  (f: org.apache.spark.TaskContext => Unit)org.apache.spark.TaskContext <and>
  (listener: org.apache.spark.util.TaskCompletionListener)org.apache.spark.TaskContext
 does not take type parameters
[ERROR]       taskContext.addTaskCompletionListener[Unit]((_: TaskContext) =>
[ERROR]                   ^
[ERROR] one error found
```

Reply:
I think we can use the old API for spark 2.x and the new API for spark 3.0.

Reply:
How do you suggest we select between the APIs when using different versions of Spark? We can't do this at runtime since the issue here is a compile time issue. We could cross-compile for Scala 2.11 and 2.12 (doing something like this: https://blog.ssanj.net/posts/2017-06-26-crosscompile-multiple-versions-of-scala-via-sbt.html) but I'm not sure how to do this with Maven - is there any prior art in the Scala world which might serve as inspiration?

Reply:
I agree with you that Scala really wasn't meant for this conditional compilation. I think the easiest way probably is to create a branch to support spark 2.3 and prior versions, with your PR the master branch can support spark 2.4+ versions.

Reply:
A branch `spark-2.3` is created from current `master` branch, I think this PR can merge into master now. @junshi15

Reply:
LGTM